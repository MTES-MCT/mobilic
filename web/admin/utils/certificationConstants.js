/**
 * Certification constants
 * Ensures consistency between backend (English) and frontend (French) naming
 */

// TODO: After PR merge, uncomment these imports
// import {
//   CertificationBronze,
//   CertificationArgent,
//   CertificationOr,
//   CertificationDiamant
// } from "common/utils/icons";

const MEDAL_LEVELS_CONFIG = {
  BRONZE: { frontend: "BRONZE", display: "Bronze", filename: "bronze" },
  SILVER: { frontend: "ARGENT", display: "Argent", filename: "argent" },
  GOLD: { frontend: "OR", display: "Or", filename: "or" },
  DIAMOND: { frontend: "DIAMANT", display: "Diamant", filename: "diamant" }
};

export const BACKEND_MEDAL_LEVELS = Object.keys(MEDAL_LEVELS_CONFIG);
export const FRONTEND_MEDAL_LEVELS = Object.values(MEDAL_LEVELS_CONFIG).map(
  config => config.frontend
);

export const BACKEND_TO_FRONTEND_MAPPING = Object.entries(
  MEDAL_LEVELS_CONFIG
).reduce(
  (acc, [backend, config]) => ({ ...acc, [backend]: config.frontend }),
  {}
);

export const MEDAL_DISPLAY_LABELS = Object.entries(MEDAL_LEVELS_CONFIG).reduce(
  (acc, [, config]) => ({ ...acc, [config.frontend]: config.display }),
  {}
);

export const MEDAL_CDN_FILENAMES = Object.entries(MEDAL_LEVELS_CONFIG).reduce(
  (acc, [, config]) => ({ ...acc, [config.frontend]: config.filename }),
  {}
);

export const CDN_BASE_URL =
  "https://cdn.jsdelivr.net/gh/MTES-MCT/mobilic@master/common/assets/images/certification/squared";

// TODO: After PR merge, uncomment this mapping
// export const MEDAL_BADGE_COMPONENTS = {
//   BRONZE: CertificationBronze,
//   ARGENT: CertificationArgent,
//   OR: CertificationOr,
//   DIAMANT: CertificationDiamant,
//   SILVER: CertificationArgent,
//   GOLD: CertificationOr,
//   DIAMOND: CertificationDiamant
// };

/**
 * Internal utility to map a level to a specific mapping object
 * @param {string} level - Medal level (backend or frontend)
 * @param {Object} mapping - Target mapping object
 * @returns {string|null} Mapped value or null if level is invalid
 */
function getMappedValue(level, mapping) {
  if (!level) return null;
  const frontendLevel = BACKEND_TO_FRONTEND_MAPPING[level] || level;
  return mapping[frontendLevel] || null;
}

/**
 * Gets the display label for a medal level
 * @param {string} level - Medal level (backend or frontend)
 * @returns {string|null} Display label (Bronze, Argent, Or, Diamant) or null
 */
export function getMedalDisplayLabel(level) {
  return getMappedValue(level, MEDAL_DISPLAY_LABELS);
}

/**
 * Gets the CDN filename for a medal level
 * @param {string} level - Medal level (backend or frontend)
 * @returns {string|null} Filename (bronze, argent, or, diamant) or null
 */
export function getMedalCdnFilename(level) {
  return getMappedValue(level, MEDAL_CDN_FILENAMES);
}

/**
 * Gets the complete CDN URL for a medal badge
 * @param {string} level - Medal level (backend or frontend)
 * @returns {string|null} Complete CDN URL or null if level is invalid
 */
export function getMedalCdnUrl(level) {
  const filename = getMedalCdnFilename(level);
  if (!filename) return null;
  return `${CDN_BASE_URL}/${filename}.png`;
}

/**
 * Validates if a medal level is valid (backend or frontend)
 * @param {string} level - Medal level to validate
 * @returns {boolean} True if valid, false otherwise
 */
export function isValidMedalLevel(level) {
  return (
    BACKEND_MEDAL_LEVELS.includes(level) ||
    FRONTEND_MEDAL_LEVELS.includes(level)
  );
}
