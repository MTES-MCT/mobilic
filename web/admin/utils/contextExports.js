import React, { createContext, useContext } from "react";
import { useExports as useExportsHook } from "../hooks/useExports";

const ExportsContext = createContext(null);

export const useExportsContext = () => useContext(ExportsContext);

export function ExportsProvider({ children }) {
  return (
    <ExportsContext.Provider value={{ ...useExportsHook() }}>
      {children}
    </ExportsContext.Provider>
  );
}
