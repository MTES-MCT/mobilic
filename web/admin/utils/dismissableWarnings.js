export const DISMISSABLE_WARNINGS = {
  EMPLOYEE_MISSION_VALIDATION: "employee-validation",
  ADMIN_MISSION_MODIFICATION: "admin-mission-modification"
};
