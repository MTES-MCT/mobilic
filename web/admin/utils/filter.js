export function useFilter(elements, setElements) {
  function handleFilterChange(event, value) {
    const selectedIds = value.map(c => c.id);
    setElements(
      elements.map(c => ({
        ...c,
        selected: selectedIds.includes(c.id)
      }))
    );
  }

  const selectedElements = elements.filter(e => e.selected);
  return [selectedElements, handleFilterChange];
}
