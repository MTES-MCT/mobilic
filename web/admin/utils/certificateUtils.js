import { getMedalDisplayLabel, getMedalCdnUrl } from "./certificationConstants";

// ===== FONCTIONS D'INTERFACE PUBLIQUE =====
// Ces fonctions maintiennent la compatibilit√© avec l'API existante

/**
 * @deprecated Utiliser getMedalDisplayLabel de certificationConstants
 */
export function getCertificateLevelLabel(level) {
  return getMedalDisplayLabel(level);
}

/**
 * @deprecated Utiliser getMedalCdnUrl de certificationConstants
 */
export function getBadgeCdnUrl(level) {
  return getMedalCdnUrl(level);
}

/**
 * @deprecated Will be implemented after PR merge
 */
export function getCertificateBadgeComponent() {
  return null;
}

export function generateEmbedCodes(level) {
  const badgeUrl = getBadgeCdnUrl(level);
  const levelLabel = getCertificateLevelLabel(level);

  if (!badgeUrl) return { iframe: "", image: "", script: "" };

  const altText = `Certificat Mobilic ${levelLabel || level}`;
  const titleText = `Certificat Mobilic ${levelLabel || level}`;

  return {
    iframe: `<iframe src="${badgeUrl}" width="150" height="150" frameborder="0" style="border: none;" title="${titleText}"></iframe>`,

    image: `<img src="${badgeUrl}" alt="${altText}" width="150" height="150" />`,

    script: `<a href="https://mobilic.beta.gouv.fr" target="_blank" rel="noopener">
  <img src="${badgeUrl}" alt="${altText}" width="150" height="150" />
</a>`
  };
}
