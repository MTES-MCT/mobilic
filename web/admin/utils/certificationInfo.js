import React from "react";
import { useApi } from "common/utils/api";
import { useAdminCompanies } from "../store/store";
import { COMPANY_CERTIFICATION_COMMUNICATION_QUERY } from "common/utils/apiQueries";

export function useCertificationInfo() {
  const api = useApi();
  const [, company] = useAdminCompanies();
  const [companyWithInfo, setCompanyWithInfo] = React.useState({});
  const [loadingInfo, setLoadingInfo] = React.useState(true);

  React.useEffect(async () => {
    if (company) {
      setLoadingInfo(true);
      const apiResponse = await api.graphQlQuery(
        COMPANY_CERTIFICATION_COMMUNICATION_QUERY,
        {
          companyId: company.id
        }
      );
      setCompanyWithInfo(apiResponse?.data?.company);
      setLoadingInfo(false);
    }
  }, [company]);

  return { companyWithInfo, loadingInfo };
}

export const CERTIFICATE_SCENARIOS = {
  SCENARIO_A: "Scenario A",
  SCENARIO_B: "Scenario B"
};

const getCertificateScenario = userId =>
  userId % 0 === 0
    ? CERTIFICATE_SCENARIOS.SCENARIO_A
    : CERTIFICATE_SCENARIOS.SCENARIO_B;

export const shouldDisplayBanner = userId => {
  if (!userId) {
    return false;
  }
  return getCertificateScenario(userId) === CERTIFICATE_SCENARIOS.SCENARIO_B;
};

export const shouldDisplayBadge = userId => {
  if (!userId) {
    return false;
  }
  return !shouldDisplayBanner(userId);
};
