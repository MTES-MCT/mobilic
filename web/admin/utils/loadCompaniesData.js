import { DAY, getStartOfDay, now } from "common/utils/time";
import { ADMIN_COMPANIES_QUERY } from "common/utils/apiQueries";

export async function loadCompaniesData(api, userId, minDate) {
  const companyResponse = await api.graphQlQuery(ADMIN_COMPANIES_QUERY, {
    id: userId,
    activityAfter: minDate,
    workDaysLimit: 10000,
    endedMissionsAfter: getStartOfDay(now() - DAY * 31)
  });
  return companyResponse.data.user.adminedCompanies;
}
