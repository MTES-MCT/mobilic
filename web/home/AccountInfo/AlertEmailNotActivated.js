import AlertTitle from "@material-ui/lab/AlertTitle";
import Typography from "@material-ui/core/Typography";
import List from "@material-ui/core/List";
import { ListItem, ListItemIcon, ListItemText } from "@material-ui/core";
import Button from "@material-ui/core/Button";
import Alert from "@material-ui/lab/Alert";
import React from "react";
import { RESEND_ACTIVATION_EMAIL } from "common/utils/apiQueries";
import { formatApiError } from "common/utils/errors";
import { useSnackbarAlerts } from "../../common/Snackbar";
import { useApi } from "common/utils/api";
import makeStyles from "@material-ui/core/styles/makeStyles";

export default function AlertEmailNotActivated({ email }) {
  const useStyles = makeStyles(theme => ({
    inactiveAlert: {
      marginTop: theme.spacing(2)
    },
    activationIntroLabel: {
      paddingBottom: theme.spacing(2)
    },
    emailCheckIcon: {
      color: "rgb(85, 21, 15)",
      minWidth: theme.spacing(5)
    },
    emailCheckItem: {
      alignItems: "baseline"
    }
  }));

  const api = useApi();
  const alerts = useSnackbarAlerts();
  const classes = useStyles();

  async function resendActivationEmail() {
    try {
      await api.graphQlMutate(
        RESEND_ACTIVATION_EMAIL,
        { email: email },
        { context: { nonPublicApi: true } }
      );
      alerts.success("Le mail a Ã©tÃ© renvoyÃ© avec succÃ¨s", email, 6000);
    } catch (err) {
      alerts.error(formatApiError(err), email, 6000);
    }
  }

  return (
    <Alert key={0} severity="error" className={classes.inactiveAlert}>
      <AlertTitle className="bold">Adresse email non activÃ©e</AlertTitle>
      <Typography className={classes.activationIntroLabel}>
        Un email d'activation vous a Ã©tÃ© envoyÃ© Ã  l'adresse ci-dessus.
        L'activation est nÃ©cessaire pour accÃ©der aux services Mobilic.
      </Typography>
      <List disablePadding>
        <ListItem disableGutters dense className={classes.emailCheckItem}>
          <ListItemIcon className={classes.emailCheckIcon}>ðŸ‘‰</ListItemIcon>
          <ListItemText primary="Il peut mettre quelques minutes Ã  arriver, merci pour votre patience" />
        </ListItem>
        <ListItem disableGutters dense className={classes.emailCheckItem}>
          <ListItemIcon className={classes.emailCheckIcon}>ðŸ‘‰</ListItemIcon>
          <ListItemText primary="Merci de vÃ©rifier vos spams" />
        </ListItem>
        <ListItem disableGutters dense className={classes.emailCheckItem}>
          <ListItemIcon className={classes.emailCheckIcon}>ðŸ‘‰</ListItemIcon>
          <ListItemText primary="Pour recevoir nos emails sans encombres, vous pouvez ajouter mobilic@beta.gouv.fr Ã  votre liste de contacts" />
        </ListItem>
        <ListItem disableGutters dense className={classes.emailCheckItem}>
          <ListItemIcon className={classes.emailCheckIcon}>ðŸ‘‰</ListItemIcon>
          <ListItemText primary="Si vous n'avez pas reÃ§u l'email de confirmation au bout d'une demi-heure, vous pouvez le renvoyer en cliquant sur le bouton ci dessous :" />
        </ListItem>
        <ListItem disableGutters dense className={classes.emailCheckItem}>
          <ListItemIcon className={classes.emailCheckIcon} />
          <Button
            size="small"
            color="primary"
            variant="contained"
            data-qa="resendActivationEmailButton"
            onClick={resendActivationEmail}
          >
            Renvoyer l'email d'activation
          </Button>
        </ListItem>
      </List>
    </Alert>
  );
}
