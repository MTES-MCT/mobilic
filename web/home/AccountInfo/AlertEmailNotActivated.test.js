import React from "react";
import { shallow } from "enzyme";
import * as api from "common/utils/api";
import { RESEND_ACTIVATION_EMAIL } from "common/utils/apiQueries";
import AlertEmailNotActivated from "./AlertEmailNotActivated";

let graphQlSpy;

describe("AlertEmailNotActivated", () => {
  beforeAll(() => {
    const apiSpy = jest.spyOn(api, "useApi");
    apiSpy.mockReturnValue({
      graphQlMutate: jest.fn()
    });
    graphQlSpy = jest.spyOn(api.useApi(), "graphQlMutate");
  });

  const setup = propsOverwrite => {
    return shallow(<AlertEmailNotActivated {...propsOverwrite} />);
  };

  it("should render the AlertEmailNotActivated correctly ", () => {
    const customProps = { email: "test1@test.fr" };
    const wrapper = setup(customProps);
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it("should call the resend email action if the resend button is clicked", () => {
    const customProps = { email: "test2@test.fr" };
    const wrapper = setup(customProps);

    const resendEmailButton = wrapper.find(
      '[data-qa="resendActivationEmailButton"]'
    );

    resendEmailButton.props().onClick();

    expect(graphQlSpy).toHaveBeenCalledWith(
      RESEND_ACTIVATION_EMAIL,
      { email: "test2@test.fr" },
      { context: { nonPublicApi: true } }
    );
  });
});
