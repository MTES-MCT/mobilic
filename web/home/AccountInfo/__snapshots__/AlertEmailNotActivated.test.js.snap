// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlertEmailNotActivated should render the AlertEmailNotActivated correctly  1`] = `
"<WithStyles(ForwardRef(Alert)) severity=\\"error\\" className=\\"makeStyles-inactiveAlert-1\\">
  <WithStyles(ForwardRef(AlertTitle)) className=\\"bold\\">
    Adresse email non activÃ©e
  </WithStyles(ForwardRef(AlertTitle))>
  <WithStyles(ForwardRef(Typography)) className=\\"makeStyles-activationIntroLabel-2\\">
    Un email d&#39;activation vous a Ã©tÃ© envoyÃ© Ã  l&#39;adresse ci-dessus. L&#39;activation est nÃ©cessaire pour accÃ©der aux services Mobilic.
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(List)) disablePadding={true}>
    <WithStyles(ForwardRef(ListItem)) disableGutters={true} dense={true} className=\\"makeStyles-emailCheckItem-4\\">
      <WithStyles(ForwardRef(ListItemIcon)) className=\\"makeStyles-emailCheckIcon-3\\">
        ğŸ‘‰
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText)) primary=\\"Il peut mettre quelques minutes Ã  arriver, merci pour votre patience\\" />
    </WithStyles(ForwardRef(ListItem))>
    <WithStyles(ForwardRef(ListItem)) disableGutters={true} dense={true} className=\\"makeStyles-emailCheckItem-4\\">
      <WithStyles(ForwardRef(ListItemIcon)) className=\\"makeStyles-emailCheckIcon-3\\">
        ğŸ‘‰
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText)) primary=\\"Merci de vÃ©rifier vos spams\\" />
    </WithStyles(ForwardRef(ListItem))>
    <WithStyles(ForwardRef(ListItem)) disableGutters={true} dense={true} className=\\"makeStyles-emailCheckItem-4\\">
      <WithStyles(ForwardRef(ListItemIcon)) className=\\"makeStyles-emailCheckIcon-3\\">
        ğŸ‘‰
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText)) primary=\\"Pour recevoir nos emails sans encombre, vous pouvez ajouter mobilic@beta.gouv.fr Ã  votre liste de contacts\\" />
    </WithStyles(ForwardRef(ListItem))>
    <WithStyles(ForwardRef(ListItem)) disableGutters={true} dense={true} className=\\"makeStyles-emailCheckItem-4\\">
      <WithStyles(ForwardRef(ListItemIcon)) className=\\"makeStyles-emailCheckIcon-3\\">
        ğŸ‘‰
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText)) primary=\\"Si vous n\\\\'avez pas reÃ§u l\\\\'email de confirmation au bout d\\\\'une demi-heure, vous pouvez le renvoyer en cliquant sur le bouton ci dessous :\\" />
    </WithStyles(ForwardRef(ListItem))>
    <WithStyles(ForwardRef(ListItem)) disableGutters={true} dense={true} className=\\"makeStyles-emailCheckItem-4\\">
      <WithStyles(ForwardRef(ListItemIcon)) className=\\"makeStyles-emailCheckIcon-3\\" />
      <WithStyles(ForwardRef(Button)) size=\\"small\\" color=\\"primary\\" variant=\\"contained\\" data-qa=\\"resendActivationEmailButton\\" onClick={[Function: resendActivationEmail]}>
        Renvoyer l&#39;email d&#39;activation
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(ListItem))>
  </WithStyles(ForwardRef(List))>
</WithStyles(ForwardRef(Alert))>"
`;
